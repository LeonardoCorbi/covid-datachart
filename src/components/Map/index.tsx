import { Dispatch, SetStateAction, useCallback } from 'react';
import { ComposableMap, Geographies, Geography } from 'react-simple-maps';
import { ICountryData, IQueryParams } from '../../../pages';
import { COLOR_RANGES, GEOGRAPHY } from '../../constants';
import { IFindByDate, MethodType } from '../../services/Cases/dtos/requests/findByDate.request';

interface IMap {
  tooltip: string;
  countryData: ICountryData[];
  method: MethodType;
  setTooltip: Dispatch<SetStateAction<string>>;
  setShowToolTip: Dispatch<SetStateAction<boolean>>;
}

const Map = ({
  method,
  tooltip,
  setTooltip,
  countryData,
  setShowToolTip,
}: IMap) => {
  const getCountryColor = useCallback((country: string) => {
    const countryVariants = countryData
      .find(({ name }) => name === country)?.variants;

    const count = Object
      .values(countryVariants ?? {})
      .reduce((prev, curr) => prev + curr, 0);
    
    return COLOR_RANGES.find(({ range }) => {
      const rangeNumber = method === 'total' ? range[1] : range[0];
      return count >= rangeNumber;
    })?.color;
  }, [countryData, method]);

  const onMouseEnter = useCallback((countryName: string) => {
    setTooltip(countryName);
    setShowToolTip(true);
  }, [setShowToolTip, setTooltip]);

  const onMouseLeave = useCallback(() => {
    setShowToolTip(false);
  }, [setShowToolTip]);

  return (
    <ComposableMap>
      <Geographies data-tip={tooltip} geography={GEOGRAPHY}>
        {
          ({ geographies }) => geographies.map((geo) => {
            const countryName = geo.properties.NAME_LONG;
            const commonStyle = {
              strokeWidth: .2,
              outline: 'none',
              fill: getCountryColor(countryName),
            };
    
            const countryStyle = {
              default: { ...commonStyle, opacity: .9 },
              hover: { ...commonStyle, strokeWidth: 1 },
              pressed: { ...commonStyle, strokeWidth: 1 },
            };
            return (
              <Geography
                stroke="black"
                geography={geo}
                key={geo.rsmKey}
                strokeWidth={.15}
                data-tip={tooltip}
                style={countryStyle}
                onMouseLeave={onMouseLeave}
                onMouseEnter={() => onMouseEnter(countryName)}
              />
            );
          })
        }
      </Geographies>
    </ComposableMap>
  );
};

export default Map;
