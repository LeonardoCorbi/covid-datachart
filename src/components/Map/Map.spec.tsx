import { render, fireEvent } from '@testing-library/react';
import renderer from 'react-test-renderer';
import Map from '.';

const component = () => {
  const cb = jest.fn();
  return (
    <Map
      tooltip='Teste'
      method='partial'
      countryData={[]}
      setTooltip={cb}
      setShowToolTip={cb}
    />
  );
};

describe('Map component', () => {
  it('should render correctly', () => {
    const { getAllByTestId } = render(component());

    const tree = renderer.create(component()).toJSON();
    expect(tree).toMatchSnapshot();

    const [country] = getAllByTestId('country');
    
    expect(country).toBeDefined();
  });

  it('should have hover action', () => {
    const { getAllByTestId } = render(component());
    const countries = getAllByTestId('country');
    
    countries.forEach((country) => {
      fireEvent.mouseEnter(country);

      expect(<h2 />).toBeDefined();
    });
  });
});
