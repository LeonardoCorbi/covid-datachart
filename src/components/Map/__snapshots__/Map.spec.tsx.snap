// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Map component should render correctly 1`] = `
<svg
  className="rsm-svg "
  data-testid="mapContainer"
  style={
    Object {
      "transform": "translateX(-4%)",
    }
  }
  viewBox="0 0 800 600"
>
  <g
    className="rsm-geographies "
    data-tip="Teste"
  />
</svg>
`;
